### PR TITLE
DO NOT MERGE YET Stream Tutorial 2

### PR DESCRIPTION
**Pull request by acogoluegnes**

### CODE REVIEW

File: tutorials/tutorial-two-java-stream.md [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+<TutorialsIntro/>
+
+## Offset Tracking
+
+### Setup
+
+In this part of the tutorial we'll write two programs in Java; a producer that sends a wave of messages with a poison message at the end, and a consumer that receives messages and stops when it gets the poison message.
+We'll see how a consumer can navigate through a stream and can even restart where it left off in a previous execution.
+
+We'll use [the stream Java client](/tutorials/tutorial-one-java-stream#using-the-java-stream-client).
```
[the remark this answers is not in the document]
Comment (acogoluegnes) [reply tutorials/tutorial-two-java-stream.md:39]: Sounds good, I'll update the tutorial for consistency's sake.

**Review by lukebakken — COMMENTED** [review]
What do you think about using the word "marker" instead of "poison" to indicate stopping the consumer? The only reason I bring it up is that "poison" is used in our docs to describe messages that cause consumers to throw exceptions or otherwise have errors when processing.

I can make the change, too.

**Review by lukebakken — APPROVED** [review]
I replaced `poison` with `marker` 😇 Tutorial code updated here as well - https://github.com/rabbitmq/rabbitmq-tutorials/pull/421

**Review by binarin — COMMENTED** [review]
I've followed the tutorial from the start to the end, everyting is clear and works exactly as it says. It's just a bit bumpy if you want to follow it from scratch, as there is no easy way for getting all the boilerplate ready, and some pieces of code are omitted from the text of the tutorial.

File: tutorials/tutorial-two-java-stream.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+## Offset Tracking
+
+### Setup
+
+This part of the tutorial consists in writing two programs in Java; a producer that sends a wave of messages with a marker message at the end, and a consumer that receives messages and stops when it gets the marker message.
+It shows how a consumer can navigate through a stream and can even restart where it left off in a previous execution.
+
+This tutorial uses [the stream Java client](/tutorials/tutorial-one-java-stream#using-the-java-stream-client).
+Make sure to follow [the setup steps](/tutorials/tutorial-one-java-stream#setup) from the first tutorial.
```
[the remark this answers is not in the document]
Comment (acogoluegnes) [reply tutorials/tutorial-two-java-stream.md:40]: I don't think it's manageable: we have dozens of tutorials, it's very unlikely we manage to maintain all of them with the start and end states.

File: tutorials/tutorial-two-java-stream.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+IntStream.range(0, messageCount).forEach(i -> {
+    String body = i == messageCount - 1 ? "marker" : "hello";
+    producer.send(producer.messageBuilder()
+                          .addData(body.getBytes(UTF_8))
+                          .build(),
+                  ctx -> confirmedLatch.countDown()
+    );
+});
+
+boolean completed = confirmedLatch.await(60, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (acogoluegnes) [reply tutorials/tutorial-two-java-stream.md:84]: That's right, we should mention in the first tutorial that one should wait for the publish confirm, I'll open a new PR for this.

[the remark this answers is not in the document]
Comment (acogoluegnes) [reply tutorials/tutorial-two-java-stream.md:84]: https://github.com/rabbitmq/rabbitmq-website/pull/1963

File: tutorials/tutorial-two-go-stream.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+The routine deals with the messages and sends `true` to the second channel when the expected number of confirmations is reached:
+
+```go
+func handlePublishConfirm(confirms stream.ChannelPublishConfirm, messageCount int, ch chan bool) {
+	go func() {
+		confirmedCount := 0
+		for confirmed := range confirms {
+			for _, _ = range confirmed {
```
Comment (Gsantomaggio) [inline tutorials/tutorial-two-go-stream.md:118]: Here, we count the confirmations,  but:
```golang
	for _, msg := range confirmed {
		 if msg.IsConfirmed() {
		
```

With `.IsConfirmed()`, you are sure that the server has confirmed the message. 

not sure we have to deal with `.IsConfirmed()` here or maybe directly to another tutorial.

File: tutorials/tutorial-two-go-stream.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+var firstOffset int64 = -1
+var lastOffset atomic.Int64
+ch := make(chan bool)
+messagesHandler := func(consumerContext stream.ConsumerContext, message *amqp.Message) {
+    if atomic.CompareAndSwapInt64(&firstOffset, -1, consumerContext.Consumer.GetOffset()) {
+        fmt.Println("First message received.")
+    }
+    if string(message.GetData()) == "marker" {
+        lastOffset.Store(consumerContext.Consumer.GetOffset())
+        consumerContext.Consumer.Close()
```
Comment (Gsantomaggio) [inline tutorials/tutorial-two-go-stream.md:156]: I usually suggest closing the consumer outside the message handler.

File: tutorials/tutorial-two-go-stream.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    }
+}
+
+offsetSpecification := stream.OffsetSpecification{}.First()
+_, _ = env.NewConsumer(streamName, messagesHandler,
+    stream.NewConsumerOptions().
+        SetOffset(offsetSpecification))
+fmt.Println("Started consuming...")
+_ = <-ch
+
```
Comment (Gsantomaggio) [inline tutorials/tutorial-two-go-stream.md:167]: Here you can declare the `consumer, err := .....` then:

```golang
_ = <-ch
consumer.close()
```

**Review by Gsantomaggio — CHANGES_REQUESTED** [review]

File: tutorials/tutorial-two-go-stream.md (hunk 5 above)
Comment (Gsantomaggio) [reply tutorials/tutorial-two-go-stream.md]: done via: https://github.com/rabbitmq/rabbitmq-website/pull/1952/commits/81ddb9e4699a7701c026571ba99fc11e091c6837

**Pull request merged**